src/lib/Telemetry/telemetry.cpp

```diff
@@ -190,120 +190,77 @@ bool Telemetry::RXhandleUARTin(uint8_t data)
+            CRSF::crsf2msp.parse(package);
+        }
+        //fall through: if no TCP client we just want to forward MSP over the link
+    }
+    #endif
 
+    //handle OTA
     uint8_t targetIndex = 0;
     bool targetFound = false;
 
-
-    if (header->type >= CRSF_FRAMETYPE_DEVICE_PING)
+    //store crsf telemetry in 1-slot LIFO buffer, overwrite exising data (if the crsf package fits in the buffer)
+    for (int8_t i = 0; i < payloadTypesCount - 2; i++)
     {
-        const crsf_ext_header_t *extHeader = (crsf_ext_header_t *) package;
-
-        if (header->type == CRSF_FRAMETYPE_ARDUPILOT_RESP)
+        if (header->type == payloadTypes[i].type && CRSF_FRAME_SIZE(header->frame_size) <= payloadTypes[i].size)
         {
-            // reserve last slot for adrupilot custom frame with the sub type status text: this is needed to make sure the important status messages are not lost
-            if (package[CRSF_TELEMETRY_TYPE_INDEX + 1] == CRSF_AP_CUSTOM_TELEM_STATUS_TEXT)
-            {
-                targetIndex = payloadTypesCount - 1;
-            }
-            else
-            {
-                targetIndex = payloadTypesCount - 2;
-            }
-            targetFound = true;
-        }
-        else if (extHeader->orig_addr == CRSF_ADDRESS_FLIGHT_CONTROLLER)
-        {
-            targetIndex = payloadTypesCount - 2;
-            targetFound = true;
-
-            #if defined(USE_MSP_WIFI) && defined(TARGET_RX)
-                // this probably needs refactoring in the future, I think we should have this telemetry class inside the crsf module
-                if (wifi2tcp.hasClient() && (header->type == CRSF_FRAMETYPE_MSP_RESP || header->type == CRSF_FRAMETYPE_MSP_REQ)) // if we have a client we probs wanna talk to it
-                {
-                    DBGLN("Got MSP frame, forwarding to client, len: %d", currentTelemetryByte);
-                    CRSF::crsf2msp.parse(package);
-                }
-                else // if no TCP client we just want to forward MSP over the link
-            #endif
-            {
-                // larger msp resonses are sent in two chunks so special handling is needed so both get sent
-                if (header->type == CRSF_FRAMETYPE_MSP_RESP)
-                {
-                    // there is already another response stored
-                    if (payloadTypes[targetIndex].updated)
-                    {
-                        // use other slot
-                        targetIndex = payloadTypesCount - 1;
-                    }
-
-                    // if both slots are taked do not overwrite other data since the first chunk would be lost
-                    if (payloadTypes[targetIndex].updated)
-                    {
-                        targetFound = false;
-                    }
-                }
-            }
-        }
-        else
-        {
-            targetIndex = payloadTypesCount - 1;
+            targetIndex = i;
             targetFound = true;
```

Comment:
there should probably be a break after this line to stop the loop early.

Reply:
"break;" added